### PR TITLE
For issue #2459 Adding social meta tags to display EBWiki content cor…

### PR DESCRIPTION
**Pull request by campgurus**
This PR is aimed at getting EBWiki content to display correctly on Twitter and Facebook et al.

With these changes, We can probably also remove a gem dependency : 'metamagic'

Facebook and Twitter provide resoruces for testing our urls, but I can't use them from localhost.

As soon as we put this on staging, we can test the urls at these resources to see how they disply:

* https://cards-dev.twitter.com/validator
* https://developers.facebook.com/tools/debug/

In your PR did you:

  - [x ] Include a description of the changes?
  - [x ] Mention the issue the PR addresses?
  - [ ] Include screenshots of any changes to the UI?
  - [x] Isolate any changes to gems (meaning that any new, updated, or removed gems and resulting code changes should be in their own PR)?
  - [ ] Add and/or update specs for your code?

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -32,10 +32,13 @@ def user_for_paper_trail
     user_signed_in? ? current_user.id : 'Guest'
   end
 
+  def default_url_options
+    # need heroku config:set HOST=www.ebwiki.org
+    { host: ENV['HOST'] || 'localhost:3000' }
```
Comment (trystant) [inline app/controllers/application_controller.rb]: Let's put the second string in a value in `config/locales/en.yml`, perhaps in a value named `en.development.host` ?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -32,10 +32,13 @@ def user_for_paper_trail
...
+  end
+
   private
- 
 
-  
   def log_invalid_token_attempt
     warning_message = 'Invalid Auth Token error'
```
Comment (trystant) [inline app/controllers/application_controller.rb]: This static message could also be put in `config/locales/en.yml`

File: config/meta.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+meta_product_name: "End Bias Wiki"
+meta_title: "EBWiki - Ending Violence Against People of Color"
+meta_description: "EBWiki is a new web application working to harness the power of community to end bias in law enforcement that has led to disparate treatment of people of color with tragic results."
+meta_image: "EBWikiHeadLogo.png"     # should exist in `app/assets/images/`
+meta_keywords: "police killing abuse"
```
Comment (trystant) [inline config/meta.yml]: We can add some additional keywords here, can't we?

File: app/helpers/meta_tags_helper.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
+  # frozen_string_literal: true
+
+  # Helper for meta tags to assign default tags or page-specific content_tag
+
+  def meta_title
+    content_for?(:meta_title) ? content_for(:meta_title) : DEFAULT_META['meta_title']
```
Comment (trystant) [inline app/helpers/meta_tags_helper.rb]: I think this can also be expressed as:
`content_for?(:meta_title) || DEFAULT_META['meta_title']`

Comment (campgurus) [reply app/helpers/meta_tags_helper.rb]: Just saw these. Will implement

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (campgurus) [reply app/controllers/application_controller.rb]: do you mean adding this to en.yml:

```
  development:
    host: 'localhost:3000'
```
and adjusting the method to pul from that file?

File: config/meta.yml (hunk 3 above)
Comment (campgurus) [reply config/meta.yml]: yes. open to suggestions. they will be easy to add

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (campgurus) [reply app/controllers/application_controller.rb]: i Added this to the yaml file, but can't actually  pull the value into the controller,

File: app/helpers/meta_tags_helper.rb (hunk 4 above)
Comment (campgurus) [reply app/helpers/meta_tags_helper.rb]: this is done.

**Comment by campgurus** [comment]
I started the tutorial again and get the same errors as before. These errors occur when I create the default_url_options method in the application_controller.rb.  

```
def default_url_options
  { host: ENV["HOST"] || "localhost:3000" }
end
```

Here is the feedback from the failing tests:

```

  1) CasesController#create when valid saves the new case & redirects to show the case created
     Failure/Error: expect(response).to redirect_to(case_path(Case.last))
     
     TypeError:
       class or module required for rescue clause
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/shoulda-matchers-3.1.2/lib/shoulda/matchers/action_controller/redirect_to_matcher.rb:82:in `rescue in redirects_to_url?'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/shoulda-matchers-3.1.2/lib/shoulda/matchers/action_controller/redirect_to_matcher.rb:77:in `redirects_to_url?'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/shoulda-matchers-3.1.2/lib/shoulda/matchers/action_controller/redirect_to_matcher.rb:68:in `matches?'
     # ./spec/controllers/cases_controller_spec.rb:88:in `block (4 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # Minitest::Assertion:
     #   Expected response to be a redirect to <http://test.host/cases/61title> but was a redirect to <http://localhost/cases/61title>.
     #   Expected "http://test.host/cases/61title" to be === "http://localhost/cases/61title".
     #   /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/minitest-5.11.3/lib/minitest/assertions.rb:139:in `assert'

  2) CasesController#update when valid success
     Failure/Error: expect(response).to redirect_to(case_path(this_case))
     
     TypeError:
       class or module required for rescue clause
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/shoulda-matchers-3.1.2/lib/shoulda/matchers/action_controller/redirect_to_matcher.rb:82:in `rescue in redirects_to_url?'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/shoulda-matchers-3.1.2/lib/shoulda/matchers/action_controller/redirect_to_matcher.rb:77:in `redirects_to_url?'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/shoulda-matchers-3.1.2/lib/shoulda/matchers/action_controller/redirect_to_matcher.rb:68:in `matches?'
     # ./spec/controllers/cases_controller_spec.rb:137:in `block (4 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # Minitest::Assertion:
     #   Expected response to be a redirect to <http://test.host/cases/63title> but was a redirect to <http://localhost/cases/63title>.
     #   Expected "http://test.host/cases/63title" to be === "http://localhost/cases/63title".
     #   /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/minitest-5.11.3/lib/minitest/assertions.rb:139:in `assert'

  3) User follows an case from show User arrives at the case show page and clicks to unfollow
     Failure/Error: click_link 'Unfollow'
     
     Capybara::ElementNotFound:
       Unable to find link "Unfollow"
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/capybara-3.7.0/lib/capybara/node/finders.rb:295:in `block in synced_resolve'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/capybara-3.7.0/lib/capybara/node/base.rb:82:in `synchronize'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/capybara-3.7.0/lib/capybara/node/finders.rb:286:in `synced_resolve'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/capybara-3.7.0/lib/capybara/node/finders.rb:33:in `find'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/capybara-3.7.0/lib/capybara/node/actions.rb:41:in `click_link'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/capybara-3.7.0/lib/capybara/session.rb:728:in `block (2 levels) in <class:Session>'
     # /Users/dariusgoore/.rvm/gems/ruby-2.5.1@ebwiki/gems/capybara-3.7.0/lib/capybara/dsl.rb:51:in `block (2 levels) in <module:DSL>'
     # ./spec/features/user_follows_case_spec.rb:20:in `block (2 levels) in <top (required)>'
```

**Comment by campgurus** [comment]
I am going to start again. closing this PR for now.

**Pull request closed without merging**